core/rfb.js

```diff
@@ -66,20 +66,50 @@ const extendedClipboardActionPeek    = 1 << 26;
 const extendedClipboardActionNotify  = 1 << 27;
 const extendedClipboardActionProvide = 1 << 28;
 
+// Properties a raw channel must have, WebSocket and RTCDataChannel are two examples
+const rawChannelProps = [
+    "send",
+    "close",
+    "binaryType",
+    "addEventListener",
+    "onerror",
+    "onmessage",
+    "onopen",
+    "protocol",
+    "readyState",
+];
 
 export default class RFB extends EventTargetMixin {
-    constructor(target, url, options) {
+    constructor(target, urlOrChannel, options) {
         if (!target) {
             throw new Error("Must specify target");
         }
-        if (!url) {
-            throw new Error("Must specify URL");
+        if (!urlOrChannel) {
+            throw new Error("Must specify URL or raw channel");
         }
 
         super();
 
         this._target = target;
-        this._url = url;
+
+        if (typeof urlOrChannel === "string") {
+            this._url = urlOrChannel;
+        } else {
+            this._url = null;
+            this._rawChannel = urlOrChannel;
+            for (let i = 0; i < rawChannelProps.length; i++) {
+                const prop = rawChannelProps[i];
+                if (urlOrChannel[prop] === undefined) {
+                    console.error('Raw channel missing property', prop);
+                    this._rawChannel = null;
+                    break;
+                }
+            }
```

Comment:
I guess it's the WebSock object that is mostly concerned about these things, so can we put the check there? That way all of the code is in the same place.

Reply:
This turned out to have unexpected repercussions. The tests weren't always using the FakeWebsocket and the FakeWebsocket didn't have some properties set to null that were expected to be.

Reply:
But I guess you get it working?

Reply:
Yep. Just wanted you to be aware of the additional changes